src/etos_suite_runner/lib/suite.py

```diff
@@ -0,0 +1,322 @@
+# Copyright 2022 Axis Communications AB.
+#
+# For a full list of individual contributors, please see the commit history.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Test suite handler."""
+import logging
+import threading
+import time
+
+from etos_lib.logging.logger import FORMAT_CONFIG
+from eiffellib.events import EiffelTestSuiteStartedEvent
+
+from .executor import Executor
+from .graphql import request_test_suite_finished, request_test_suite_started
+from .log_filter import DuplicateFilter
+
+
+class SubSuite:
+    """Handle test results and tracking of a single sub suite."""
+
+    released = False
+
+    def __init__(self, etos, environment):
+        """Initialize a sub suite."""
+        self.etos = etos
+        self.environment = environment
+        self.name = self.environment.get("name")
+        self.logger = logging.getLogger(f"SubSuite - {self.name}")
+        self.logger.addFilter(DuplicateFilter(self.logger))
+        self.test_suite_started = {}  # This is set by a different thread.
+        self.test_suite_finished = {}
+
+    @property
+    def finished(self):
+        """Whether or not this sub suite has finished."""
+        return bool(self.test_suite_finished)
+
+    @property
+    def started(self):
+        """Whether or not this sub suite has started."""
+        return bool(self.test_suite_started)
+
+    def request_finished_event(self):
+        """Request a test suite finished event for this sub suite."""
+        # Prevent ER requests if we know we're not even started.
+        if not self.started:
+            return
+        # Prevent ER requests if we know we're already finished.
+        if not self.test_suite_finished:
+            self.test_suite_finished = request_test_suite_finished(
+                self.etos, self.test_suite_started["meta"]["id"]
+            )
+
+    def outcome(self):
+        """Outcome of this sub suite.
+
+        :return: Test suite outcome from the test suite finished event.
+        :rtype: dict
+        """
+        if self.finished:
+            return self.test_suite_finished.get("testSuiteOutcome", {})
```

Comment:
This needs to get the "data" key first.
All main suites will always fail due to this not returning a proper outcome